### PR TITLE
Streamline ImplSymbolVisitors

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

**Comment by mfussenegger** [comment]
@seut @matriv you may want to take a look at this. It isn't ready yet (although I think tests pass)

But I'm not sure if this is even an improvement or just a different way of doing the same thing.

I started this for three reasons:

 1. I didn't fully understand the difference between ImplementationSymbolVisitor and CollectInputSymbolVisitor
 2. Complex class hierarchy with AbstractImplementationSymbolVisitor & BaseImplementationSymbolVisitor
 3. Not really type safe - We use generics for the docExpressions, but the context itself is not used as a generic so we lose the type safety.

Here's why I'm not sure if these changes really make sense:

 1. Could be solved easier by renaming the existing classes and adding docs.
 2. I'm not sure if this became simpler
 3. This would be fixed, although there are still a few places where I had to do it unsafe.

**Comment by seut** [comment]
In general I like it:

1. I think it's much better than just renaming and adding docs imho, but adding some docs specially to the ImplementationResolver would be great ;)
2. imho much cleaner, not sure if simpler...
3. great! not sure if the leftovers are really important as they afaik use Object as a concrete type anyway or?

**Comment by mfussenegger** [comment]
Regarding 3)

The leftovers are mostly system tables where we have `Iterable<?>` and concrete Expression implementation which are not using Object but some specific type.
E.g. sys.jobs has `Iterable<JobContext>`, and all expressions expect a JobContext.

If we changed the data structure to have the referenceResolver and the iterable in the same place this could be solved. But doing so would require some more work.

File: sql/src/main/java/io/crate/operation/collect/DocInputFactory.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Specialized InputFactory for Lucene symbols/expressions.
+ *
+ * See {@link InputFactory} for an explanation what a InputFactory does.
+ */
+public class DocInputFactory {
+
+    private final ReferenceResolver<? extends LuceneCollectorExpression<?>> referenceResolver;
+    private final InputFactory implResolver;
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/DocInputFactory.java]: maybe rename this as well (implFactory)?

File: sql/src/main/java/io/crate/operation/collect/RowShardResolver.java [hunk 2]
```diff
@@ -58,19 +56,24 @@ public RowShardResolver(Functions functions,
                             List<? extends Symbol> primaryKeySymbols,
                             @Nullable ColumnIdent clusteredByColumn,
                             @Nullable Symbol routingSymbol) {
-        Visitor visitor = new Visitor(functions);
+        InputFactory implResolver = new InputFactory(functions);
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/RowShardResolver.java]: also here

File: sql/src/main/java/io/crate/operation/collect/sources/FileCollectSource.java [hunk 3]
```diff
@@ -47,20 +48,22 @@
 public class FileCollectSource implements CollectSource {
 
     private final ClusterService clusterService;
-    private final FileCollectInputSymbolVisitor fileInputSymbolVisitor;
     private final Map<String, FileInputFactory> fileInputFactoryMap;
+    private final InputFactory implResolver;
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/sources/FileCollectSource.java]: and here

File: sql/src/main/java/io/crate/operation/collect/sources/SingleRowSource.java [hunk 4]
```diff
@@ -57,13 +59,14 @@ public SingleRowSource(Functions functions, NestedReferenceResolver clusterRefRe
...
         collectPhase = collectPhase.normalize(clusterNormalizer, null);
-        ImplementationSymbolVisitor nodeImplementationSymbolVisitor = new ImplementationSymbolVisitor(functions);
         if (collectPhase.whereClause().noMatch()) {
             return ImmutableList.<CrateCollector>of(RowsCollector.empty(downstream));
         }
         assert !collectPhase.whereClause().hasQuery()
             : "WhereClause should have been normalized to either MATCH_ALL or NO_MATCH";
-        ImplementationSymbolVisitor.Context ctx = nodeImplementationSymbolVisitor.extractImplementations(collectPhase.toCollect());
+
+        InputFactory resolver = new InputFactory(functions);
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/sources/SingleRowSource.java]: here as well

File: sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java [hunk 5]
```diff
@@ -53,13 +53,13 @@
 
     private final ClusterService clusterService;
     private final Functions functions;
-    private final TableFunctionImplementationVisitor implementationVisitor;
+    private final InputFactory implResolver;
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java]: another one ;)

File: sql/src/main/java/io/crate/operation/projectors/ProjectionToProjectorVisitor.java [hunk 6]
```diff
@@ -70,7 +77,7 @@
     private final Settings settings;
     private final TransportActionProvider transportActionProvider;
     private final BulkRetryCoordinatorPool bulkRetryCoordinatorPool;
-    private final ImplementationSymbolVisitor symbolVisitor;
+    private final InputFactory implResolver;
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/ProjectionToProjectorVisitor.java]: also here

File: sql/src/test/java/io/crate/operation/InputFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.List;
+
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+import static org.hamcrest.core.Is.is;
+
+public class InputFactoryTest extends CrateUnitTest {
+
+    private SqlExpressions expressions = new SqlExpressions(ImmutableMap.of(T3.T1, T3.TR_1), T3.TR_1);
+    private InputFactory resolver = new InputFactory(expressions.functions());
```
Comment (seut) [inline sql/src/test/java/io/crate/operation/InputFactoryTest.java]: rename as well

**Pull request merged**